src/main/java/emu/grasscutter/game/quest/QuestManager.java

```diff
@@ -227,29 +230,95 @@ public GameQuest addQuest(int questId) {
         return quest;
     }
 
-    public void startMainQuest(int mainQuestId) {
-        var mainQuestData = GameData.getMainQuestDataMap().get(mainQuestId);
+    public void startMainQuest(boolean notify, int mainQuestId, boolean useInput, String paramStr, int... params) {
+        MainQuestData mainQuestData = GameData.getMainQuestDataMap().get(mainQuestId);
+
+        if (mainQuestData == null) return; // check for bin output
+
+        // Main quest
+        GameMainQuest mainQuest = getMainQuestById(mainQuestId);
+
+        // Create main quest if it doesnt exist
+        if (mainQuest == null) {
+            mainQuest = addMainQuest(mainQuestId);
+        } else {
+            return;
+        }
+
+        int[] subQuestBegun = new int[(int) mainQuest.getChildQuests().values().stream().count()];
+
+        for (GameQuest subQuest : mainQuest.getChildQuests().values().stream().toList()) {
+            val acceptCond = subQuest.getQuestData().getAcceptCond();
+            boolean shouldAccept = false;
+            if (acceptCond.size() == 0) {
+                // generally means QUEST_COND_NONE
+                shouldAccept = true;
+            } else {
+                int[] accept = new int[acceptCond.size()];
+
+                for (int i = 0; i < acceptCond.size(); i++) {
+                    val condition = acceptCond.get(i);
+                    int[] condParam = condition.getParam();
+                    String condParamStr = condition.getParamStr();
+                    if (useInput) { // differentiate usage for queue event
+                        condParam = params;
+                        condParamStr = paramStr;
+                    }
+                    boolean result = getPlayer().getServer().getQuestSystem().triggerCondition(subQuest, condition, condParamStr, condParam);
+                    accept[i] = result ? 1 : 0;
+                }
+                shouldAccept = LogicType.calculate(subQuest.getQuestData().getAcceptCondComb(), accept);
+            }
+
+            if (shouldAccept)
+            // keep track instead of actually starting
+                subQuestBegun[subQuest.getQuestData().getOrder()-1] = 1;
+        }
 
-        if (mainQuestData == null) {
+        // if none of the subquest should start
+        if (!LogicType.calculate(LogicType.LOGIC_OR, subQuestBegun)) {
+            removeMainQuest(mainQuestId);
```

Comment:
Hmm, this also looks like a very big workaround, around the current system. Wouldn't this way of handling the quests trigger many db writes?

Reply:
this is why i have an array that keep check of the order to start rather than actually starting the GameQuest, this will not trigger the db write because none of the quest will be started and finish before this if statement. But yea i guess i shouldn't add it to the map before this if statement anyways